File: src/wtime.py
import sys
import time
import subprocess
from colorama import init, Fore, Style

init(autoreset=True)

def print_error(message: str):
    print(f"\n{Fore.RED}[ERROR]{Style.RESET_ALL} {message}")

def print_time(label: str, seconds: float):
    print(f"\n{Fore.GREEN}{label:<15}:{Style.RESET_ALL} {seconds:.6f} sec")

def calculate_time(start, returncode):
    end = time.perf_counter()

    elapsed = end - start
    print_time("Real Time", elapsed)
    sys.exit(returncode)

def main():
    if len(sys.argv) < 2:
        print_error("Usage: wtime <command>")
        sys.exit(1)

    command = sys.argv[1:]

    try:
        start = time.perf_counter()
        result = subprocess.run(command, shell=True)

        calculate_time(start, result.returncode)

    except KeyboardInterrupt:
        print_error('Operation canceled by user')
        calculate_time(start, 2)

    except Exception as e:
        print_error(f"Failed to execute command: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: src/wtime.cpp
#include <chrono>
#include <cstdlib>
#include <iostream>
#include <string>
#include <format>

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: wtime <command>\n";
        return EXIT_FAILURE;
    }
  
    std::string command;
    for (int i = 1; i < argc; ++i) {
        command += argv[i];
        if (i < argc - 1) command += ' ';
    }
  
    auto start = std::chrono::steady_clock::now();
    int result = std::system(command.c_str());
    auto end = std::chrono::steady_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
  
    std::cout << std::format("Elapsed time: {}.{:03} seconds\n",
                             duration.count() / 1000,
                             duration.count() % 1000);
    return result;
}
